fisherman-registry/src/main/java/fishermanregistry/collector/scheduled/FishCollectorScheduler.java

```
@@ -1,47 +1,49 @@
 import org.springframework.stereotype.Service;
-import org.springframework.transaction.annotation.Transactional;
 import java.time.Instant;
 import java.util.Date;
+import java.util.List;
 
 @Service
 @Slf4j
 @RequiredArgsConstructor
 public class FishCollectorScheduler {
 
-    private final CollectorRepository collectorRepository;
-    private final FishSourceClient fishSource;
+    private final FishSourceAsyncClient fishSourceWrapper;
+    private final CollectorService collectorService;
 
-    @Transactional
     @Scheduled(fixedRateString = "${fisherman-registry.collector.scheduled-rate-ms}")
     public void collect() {
         log.info("Collecting started");
 
-        var readyCollectors = collectorRepository.findReadyCollectors(Date.from(Instant.now()));
+        var readyCollectors = collectorService.getReadyCollectors();
+
+        log.info("Found {} ready collectors", readyCollectors.size());
 
         if (readyCollectors.isEmpty()) {
-            log.info("No ready collectors");
             return;
         }
 
-        log.info("Found {} ready collectors", readyCollectors.size());
-
-        var instantNow = Instant.now();
         for (var collector : readyCollectors) {
-            if (fishSource.collectFish(collector.getQuantity()).size() != collector.getQuantity()) {
-                log.info("Requesting {} fish failed", collector.getQuantity());
-                break;
-            }
-            collectorRepository.updateNextCollectionDate(
-                    Date.from(instantNow.plusMillis(collector.getInterval())), collector.getId());
+            fishSourceWrapper.collectFish(collector.getQuantity())
+                    .thenAccept(fishes -> updateNextCollectionDate(fishes, collector));
         }
+    }
 
-        log.info("Collecting executed");
+    private void updateNextCollectionDate(List<Fish> fishes, Collector collector) {
+        if (fishes.size() != collector.getQuantity()) {
```

You forgot about the tests ;). How do you think how throwing the `NotEnoughFish` exception will impact the running process? I want to see integration tests with this asynchronous behavior.

---

And besides that, I do not think it will ever be thrown because `fish-source` also throwing an exception... I think you should propose some solution for that. My proposal is that collector should collect as many fish as can and then eventually throw this exception. This probably forces changes to `fish-source` as well.

---

But if a collector collects only f.e. a half of requested quantity, then how would he know how many he is expecting during next call and would this situation be considered ass success or some recovery needed? I would have added a new field to the Collector like `debtQuantity` then. And for `fish-source` I would add a boolean flag to collect request like `bool allOrNothing`

---

About the test. It is actually tested in `FishCollectorSchedulerTest.shouldNotUpdateNextCollectDateIfNotEnoughFish()` (line 91). The exception is silently swallowed by `Async`

---

I think it will be weird if you go fishing and after two times you didn't catch a fish the third time you catch three at once ;). I know these requirements have not been defined but hey, you know the fishing domain and have some basic outlines. Why we should allow to catch more fish than the tool allows - that will be unexpected. As a developer, I want to create solutions that are easy to follow and with as little complexity as possible. By complexity, I also mean state management. Your proposal introduces a lot of complexity, and since we know the domain of fish, I think you already know that it is not needed.

Regarding integration tests - you introduce a new exception, so how it can be tested before? Am I missing something?

---

So, tool's quantity in reality defines a maximum and it is allowed to collect less fish than specified in collector and it is the expected behaviour? I was just looking at this task from a market perspective, not like an emulation of, like,  'an ocean and fishermans':)

About test. This situation was already covered in task #3, but there was just a `return` and not exception. Since exception is silently swallowed, then in integration it is not possible to catch. Because it is the last operations inside the pipeline, then probably the exception raising is overcomplication and we can return back to just `return`

---

We are creating an emulation of the fish industry - using tools that we are using at Acoustic and trying to follow best dev practices. Understanding requirements and what to do is a crucial part of or job. From what I can see, a lot of code has already disappeared from the initial proposal proposed by you in this PR, and the code still meets scraps of requirements :). In my opinion, it is a success. I'm sorry that I won't answer your question :).

Regarding test, please add a test that checks this new exception - with the assumption that you may not get enough fish, and propose some handling for that. For me, it is enough to put an error message into logs.

---

Okay, then I would stick with the current strict implementation "all or nothing". I think it is simple enough and could be easily tweaked into any side in the node where I'm throwing this exception if we have real requirements and a person to torture:)
Maybe on a diagram it would be really transparent: Requesting fish from client -> client returns some fish in a List, client implementation doesn't care about the domain, just transmits list: all or nothing; doing only transportation -> business logic how to handle the response in the scheduled task `collect` (our response is `List<Fish>`)

---

But we should still follow the fishing domain :D. So when you go fishing you collect as many fish as you can, not as many as you assumed the day before. Or maybe I don't understand fishing :D (I have never caught fish).
So following that:
client -> server returns not as many fish as expected -> log warning / error (should remove constrains on `fish-source` side)
client -> server return error (because there are no fish) -> log error

---

Okay, understood.
Btw, I remember situations when on 1 hook could be caught more than one fish if they are small enough and stars align together (probability is really small, like 1/100000, but exists) :)

---

I'll do the following: 
1) A tool has estimated maximum amount of fish that can be acquired by it (we settled on `quantity`) in a defined intervals `interval` (like we are checking the fishing rod/net once per interval)
2) Because of (1) and the domain specifics, we can get from `0` to `quantity` per tool. Then, there is no `error` - it is the expected behaviour. Warning would be strange, because logically it is rather follows uniform distribution, and then getting the `quantity` is rather awkward too. I would keep this as `info` or nothing.

---

it makes sense :)